Userland/Libraries/LibGUI/TextDocument.cpp

```diff
@@ -343,6 +364,61 @@ String TextDocument::text() const
     return builder.to_string();
 }
 
+int TextDocument::text_len_behind_cursor(TextPosition const& cursor, TextPosition const& prev_word_break) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    auto& line = this->line(cursor.line()); // lets see the text behind the cursor
+    auto line_span = Span<u32 const>(line.code_points(), line.length());
+    auto cursor_span = line_span.slice(prev_word_break.column(), cursor.column() - prev_word_break.column());
+    auto get_emoji_result = get_emoji_length_in_span(cursor_span, cursor_span.size() - 1, 1);
+    return max(get_emoji_result, 1); // do max in case of a return 0 we delete normal ascii chars
+}
+
+int TextDocument::text_len_in_front_of_cursor(TextPosition const& cursor, TextPosition const& next_word_break) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    auto& line = this->line(cursor.line()); // lets see the text behind the cursor
+    auto line_span = Span<u32 const>(line.code_points(), line.length());
+    auto cursor_span = line_span.slice(cursor.column(), next_word_break.column() - cursor.column());
+
+    // check if the entire span an emoji
+    auto emoji = Unicode::find_emoji_for_code_points(cursor_span);
+    size_t i = 0;
+
+    // Loop over span to see if we find emoji. Assume there is no emoji if no emoji is found assume there is no emoji
+    while (!emoji.has_value() && i < cursor_span.size() - 1 && cursor_span.size() > 0) {
+        auto cursor_slice = cursor_span.slice(cursor.column(), i);
+        emoji = Unicode::find_emoji_for_code_points(cursor_slice);
+        ++i;
+    }
+    if (emoji.has_value())
+        return emoji.value().code_points.size();
+
+    return 1;
+}
+
+int TextDocument::get_emoji_length_in_span(Span<u32 const> const& cursor_span, int slice_start, size_t slice_size) const
+{
+    if (cursor_span.is_empty()) {
+        return 0;
+    }
+    if (slice_start < 0) {
+        return 0; // we have gone outside of our original span and haven't found an emoji, so assume there are none in the span
+    }
+
+    auto cursor_slice = cursor_span.slice(slice_start, slice_size);
+    auto emoji = Unicode::find_emoji_for_code_points(cursor_slice);
```

Comment:
Is there a reason this fix should only apply for emoji and not any multi-codepoint grapheme cluster, such as é or ü? Or does the "emoji for code points" method handle that already?

Reply:
I'm not sure if the find_emoji_for_code_points function handles those special unicode characters. I specifically addressed the emoji aspect of the bug. I just tested it in my sandbox and it seems to delete those characters correctly (both backspace and delete). Not sure if that is because of this code or not though

Reply:
@ADKaster U+00FC and U+00E9 are both single code points :^)